### PR TITLE
[eas-cli] Use discriminated union type for BuildContext

### PR DESCRIPTION
**Pull request by wschurman**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions

This reduces the amount of casting we do of the context type.

i.e. we no longer need the `as` cast in:
```typescript
if (ctx.platform === Platform.IOS) {
    const iosContext = ctx as BuildContext<Platform.IOS>;
```

# How

Update types in `context.ts`, run tsc, fix errors.

# Test Plan

`yarn start` (runs tsc)

### CODE REVIEW

**Comment by wschurman** [comment]
/changelog-entry chore Use discriminated union type for BuildContext

File: packages/eas-cli/src/build/createContext.ts [hunk 1]
```diff
@@ -80,7 +80,7 @@ export async function createBuildContextAsync<T extends Platform>({
   };
   Analytics.logEvent(BuildEvent.BUILD_COMMAND, trackingCtx);
 
-  const commonContext: CommonContext<T> = {
+  const commonContext = {
```
Comment (dsokal) [inline packages/eas-cli/src/build/createContext.ts:83]: I'm not convinced about this change. With this line here and `as ...` below we lose type safety.

**Review by dsokal — APPROVED** [review]
Thanks, looks great!

File: packages/eas-cli/src/build/createContext.ts (hunk 1 above)
Comment (wschurman) [reply packages/eas-cli/src/build/createContext.ts:83]: This is a really good point. I did some digging and I think this PR could go quite a bit deeper. Will investigate soon, but I actually think I may abandon this effort for now (this goes a lot deeper).

**Pull request closed without merging**